### PR TITLE
[Run2_2017] Save track information

### PR DESCRIPTION
**Pull request by aperloff**
Changes to allow for the saving of track information which is stored in the MiniAOD samples. This will add a lot of information to the ntuples, so it is turned off by default. This was tested using the unit test command:

```bash
python unitTest.py test=1 run=true command="emerging=True"
```

### CODE REVIEW

File: Reflex/src/classes_def.xml [hunk 1]
```diff
@@ -1,11 +1,18 @@
...
+  <class name="std::vector<std::vector<math::XYZVector> >"/>
+  <class name="std::vector<math::XYZPoint>"/>
+  <class name="std::vector<std::vector<math::XYZPoint> >"/>
   <class name="std::vector<pat::Jet>"/>
   <class name="std::vector<std::vector<pat::Jet> >"/>
+  <class name="std::vector<std::vector<bool> >"/>
   <class name="edm::PtrVector<pat::PackedCandidate>"/>
   <class name="edm::Wrapper<std::vector<TLorentzVector> >"/>
   <class name="edm::Wrapper<std::vector<std::vector<TLorentzVector> > >"/>
+  <class name="edm::Wrapper<std::vector<std::vector<math::XYZVector> > >"/>
```
Comment (kpedro88) [inline Reflex/src/classes_def.xml:14]: were you going to remove these as redundant?

File: TreeMaker/python/makeTreeFromMiniAOD_cff.py [hunk 2]
```diff
@@ -1115,6 +1117,52 @@ def makeTreeFromMiniAOD(self,process):
...
+            'trackFilter:trksnormalizedchi2(Tracks_normalizedChi2)',
+            'trackFilter:trkspterror(Tracks_ptError)',
+            'trackFilter:trksetaerror(Tracks_etaError)',
+            'trackFilter:trksphierror(Tracks_phiError)',
+            'trackFilter:trksqoverperror(Tracks_qoverpError)',
+            'trackFilter:trksip2d(Tracks_IP2DPV0)',
+            'trackFilter:trksip2dsig(Tracks_IP2dSigPV0)',
+            'trackFilter:trksip3d(Tracks_IP3DPV0)',
+            'trackFilter:trksip3dsig(Tracks_IP3DSigPV0)',
+            'trackFilter:pfcandsdzassociatedpv(PFCandidates_dzAssociatedPV)',
```
Comment (kpedro88) [inline TreeMaker/python/makeTreeFromMiniAOD_cff.py]: This saves some quantities associated with PFCandidates (distinct from tracks), but doesn't save the PFCandidate 4-vectors. Would those also be needed? (Leading to another large space increase...)

File: TreeMaker/python/maker.py [hunk 3]
```diff
@@ -101,6 +102,8 @@ def __init__(self,parameters):
         self.VarsInt                    = cms.vstring()
         self.VarsBool                   = cms.vstring()
         self.VectorTLorentzVector       = cms.vstring()
+        self.VectorXYZVector            = cms.vstring()
+        self.VectorXYZPoint             = cms.vstring()
```
Comment (kpedro88) [inline TreeMaker/python/maker.py:108]: you should add the other new branch types here also

File: TreeMaker/python/maker.py [hunk 4]
```diff
@@ -124,6 +127,7 @@ def printSetup(self):
         print " storing hadtau variables: "+str(self.hadtau)+" w/ reclustering "+str(self.hadtaurecluster)
         print " storing Zinv variables: "+str(self.doZinv)
         print " storing semi-visible jet variables: "+str(self.semivisible)
+        print " storing emerging jet variables: "+str(self.semivisible)
```
Comment (kpedro88) [inline TreeMaker/python/maker.py]: wrong var in str()

File: Utils/src/CandidateTrackFilter.cc [hunk 5]
```diff
@@ -0,0 +1,380 @@
...
+		trks_dxypv->push_back(track.dxy());
+		trks_dxyerrorpv->push_back(track.dxyError());
+		trks_normalizedchi2->push_back(track.normalizedChi2());
+		trks_pterror->push_back(track.ptError());
+		trks_etaerror->push_back(track.etaError());
+		trks_phierror->push_back(track.phiError());
+		trks_qoverperror->push_back(track.qoverpError());
+		trks_found->push_back(track.found());
+		trks_lost->push_back(track.lost());
+		// track quality enum options: http://cmsdoxygen.web.cern.ch/cmsdoxygen/CMSSW_10_2_11_patch1/doc/html/d8/df2/classreco_1_1TrackBase.html#aeee12ec6a3ea0d65caa2695c84ab25d5
```
Comment (kpedro88) [inline Utils/src/CandidateTrackFilter.cc]: would prefer to replace doxygen links w/ github links

File: Utils/src/CandidateTrackFilter.cc [hunk 6]
```diff
@@ -0,0 +1,380 @@
...
+	infos.put(iEvent);
+	return result;
+}
+
+// ------------ other methods  ------------
+bool CandidateTrackFilter::filterOnTrack(const pat::PackedCandidate & pfCand, const reco::Track & track) const {
+	//-------------------------------------------------------------------------------------
+	// only store PFCandidate values if pt >= minPt and |eta| <= maxEta_
+	//-------------------------------------------------------------------------------------
+	if ( (track.pt() < minPt_) || (fabs(track.eta()) > maxEta_) ) return false;
```
Comment (kpedro88) [inline Utils/src/CandidateTrackFilter.cc]: use `std::abs()` rather than `fabs()`

File: Utils/src/CandidateTrackFilter.cc [hunk 7]
```diff
@@ -0,0 +1,380 @@
...
+		// fill the track values and PFCandidate values we'd like to save
+		//-------------------------------------------------------------------------------------
+		auto transientTrack = ttBuilder->build(track);
+		infos.fill(primaryVertex,lostTrack,track,transientTrack,false);
+	}
+
+	//-------------------------------------------------------------------------------------
+	// put track/PFCandidate values back into event
+	//-------------------------------------------------------------------------------------
+	bool result = infos.trks->empty();
```
Comment (kpedro88) [inline Utils/src/CandidateTrackFilter.cc]: I am concerned about letting this "filter" ever actually reject events. result should be configured as true by default, with a separate option to enable event filtering if it is desired for some reason.

File: Utils/src/CandidateTrackFilter.cc [hunk 8]
```diff
@@ -0,0 +1,380 @@
...
+bool CandidateTrackFilter::filterOnTrack(const pat::PackedCandidate & pfCand, const reco::Track & track) const {
+	//-------------------------------------------------------------------------------------
+	// only store PFCandidate values if pt >= minPt and |eta| <= maxEta_
+	//-------------------------------------------------------------------------------------
+	if ( (track.pt() < minPt_) || (fabs(track.eta()) > maxEta_) ) return false;
+
+	//-------------------------------------------------------------------------------------
+	// cut on impact parameter quantities
+	//-------------------------------------------------------------------------------------
+	if ( (std::abs(track.dz()) > maxdz_) || (std::abs(track.dxy()) > maxdxy_) ) return false;
```
Comment (kpedro88) [inline Utils/src/CandidateTrackFilter.cc]: this whole function could just be:
```C++
return track.pt() >= minPt_ and std::abs(track.eta()) <= maxEta_ and ...
```

File: Reflex/src/classes_def.xml (hunk 1 above)
Comment (aperloff) [reply Reflex/src/classes_def.xml:14]: The vector<vector<object>> weren't redundant, only the vector<object> ones.

File: TreeMaker/python/makeTreeFromMiniAOD_cff.py (hunk 2 above)
Comment (aperloff) [reply TreeMaker/python/makeTreeFromMiniAOD_cff.py]: Technically the pT of the PFCandidate doesn't need to match that of the track, but according to the MiniAOD twiki that is rare. Most often the information is the same, so the only thing you would be adding is an energy measurement. Do you think this is needed?

File: TreeMaker/python/maker.py (hunk 3 above)
Comment (aperloff) [reply TreeMaker/python/maker.py:108]: What other ones? vector<vector<object>>?

File: TreeMaker/python/maker.py (hunk 4 above)
Comment (aperloff) [reply TreeMaker/python/maker.py]: Whoops! Now fixed.

File: Utils/src/CandidateTrackFilter.cc (hunk 8 above)
Comment (aperloff) [reply Utils/src/CandidateTrackFilter.cc]: Yes, but I wanted to leave room to add other cuts/selections is need be. Putting it all into one line is not by preferred approach. Are you insisting on that format?

File: Utils/src/CandidateTrackFilter.cc (hunk 5 above)
Comment (aperloff) [reply Utils/src/CandidateTrackFilter.cc]: Done

File: Utils/src/CandidateTrackFilter.cc (hunk 6 above)
Comment (aperloff) [reply Utils/src/CandidateTrackFilter.cc]: Done

File: Utils/src/CandidateTrackFilter.cc (hunk 7 above)
Comment (aperloff) [reply Utils/src/CandidateTrackFilter.cc]: Implemented

File: TreeMaker/python/makeTreeFromMiniAOD_cff.py (hunk 2 above)
Comment (kpedro88) [reply TreeMaker/python/makeTreeFromMiniAOD_cff.py]: So the vectors starting with "Track_" and the vectors starting with "PFCandidate_" are supposed to be the same size? This breaks our naming convention.

File: TreeMaker/python/maker.py (hunk 3 above)
Comment (kpedro88) [reply TreeMaker/python/maker.py:108]: `VarsXYZVector`, etc. Every new line in `treeMaker.py`

File: Utils/src/CandidateTrackFilter.cc (hunk 8 above)
Comment (kpedro88) [reply Utils/src/CandidateTrackFilter.cc]: `if bool return bool` is a redundant pattern. You can write a single return statement on multiple lines if desired.

File: TreeMaker/python/makeTreeFromMiniAOD_cff.py (hunk 2 above)
Comment (aperloff) [reply TreeMaker/python/makeTreeFromMiniAOD_cff.py]: They should be the same size by construction. How does this break the naming convention?

Comment (kpedro88) [reply TreeMaker/python/makeTreeFromMiniAOD_cff.py]: Branches with different names are not expected to have the same size.

File: Utils/src/CandidateTrackFilter.cc (hunk 8 above)
Comment (aperloff) [reply Utils/src/CandidateTrackFilter.cc]: Okay, this has been changed.

File: TreeMaker/python/maker.py (hunk 3 above)
Comment (aperloff) [reply TreeMaker/python/maker.py:108]: Fixed

File: TreeMaker/python/makeTreeFromMiniAOD_cff.py (hunk 2 above)
Comment (aperloff) [reply TreeMaker/python/makeTreeFromMiniAOD_cff.py]: Fixed

**Comment by aperloff** [comment]
All of the comments from the review should be addressed now.

File: README.md [hunk 9]
```diff
@@ -225,6 +225,7 @@ Brief explanation of the options in [makeTree.py](./TreeMaker/python/makeTree.py
 * `doZinv`: switch to enable the Z->invisible background estimation processes (default=True)
 * `systematics`: switch to enable JEC- and JER-related systematics (default=True)
 * `semivisible`: switch to enable variables for semi-visible jets (default=True)
+* `emerging`: switch to enable variables for emerging jets (default=True)
```
Comment (kpedro88) [inline README.md]: default should be false?

File: TreeMaker/python/makeTreeFromMiniAOD_cff.py [hunk 10]
```diff
@@ -38,6 +38,8 @@ def makeTreeFromMiniAOD(self,process):
         VarsInt                    = self.VarsInt,
         VarsBool                   = self.VarsBool,
         VectorTLorentzVector       = self.VectorTLorentzVector,
+        VectorXYZVector            = self.VectorXYZVector,
+        VectorXYZPoint             = self.VectorXYZPoint,
```
Comment (kpedro88) [inline TreeMaker/python/makeTreeFromMiniAOD_cff.py:44]: for completeness the new branch lists should also be added here

File: Utils/src/CandidateTrackFilter.cc [hunk 11]
```diff
@@ -0,0 +1,366 @@
+//-----------------------------------------------------------------------------------------
```
Comment (kpedro88) [inline Utils/src/CandidateTrackFilter.cc]: this large totally irrelevant block comment can be removed

File: Utils/src/CandidateTrackFilter.cc [hunk 12]
```diff
@@ -0,0 +1,366 @@
...
+#include "DataFormats/ParticleFlowCandidate/interface/PFCandidateFwd.h"
+#include "DataFormats/PatCandidates/interface/PackedCandidate.h"
+#include "DataFormats/VertexReco/interface/VertexFwd.h"
+
+#include "TrackingTools/IPTools/interface/IPTools.h"
+#include "TrackingTools/TransientTrack/interface/TransientTrackBuilder.h"
+#include "TrackingTools/Records/interface/TransientTrackRecord.h"
+#include "MagneticField/Engine/interface/MagneticField.h"
+
+#include "CommonTools/UtilAlgos/interface/TFileService.h"
```
Comment (kpedro88) [inline Utils/src/CandidateTrackFilter.cc]: unneeded include can be removed

File: Utils/src/CandidateTrackFilter.cc [hunk 13]
```diff
@@ -0,0 +1,366 @@
...
+
+#include "Math/VectorUtil.h"
+#include "TMath.h"
+#include "TLorentzVector.h"
+
+//using namespace reco;
```
Comment (kpedro88) [inline Utils/src/CandidateTrackFilter.cc]: delete commented-out code

File: README.md (hunk 9 above)
Comment (aperloff) [reply README.md]: Done

File: TreeMaker/python/makeTreeFromMiniAOD_cff.py (hunk 10 above)
Comment (aperloff) [reply TreeMaker/python/makeTreeFromMiniAOD_cff.py:44]: Done

File: Utils/src/CandidateTrackFilter.cc (hunk 12 above)
Comment (aperloff) [reply Utils/src/CandidateTrackFilter.cc]: Done, if all you meant was the TFileService.h.

File: Utils/src/CandidateTrackFilter.cc (hunk 13 above)
Comment (aperloff) [reply Utils/src/CandidateTrackFilter.cc]: Done

File: Utils/src/CandidateTrackFilter.cc (hunk 11 above)
Comment (aperloff) [reply Utils/src/CandidateTrackFilter.cc]: Done

**Pull request merged**